packages/design-system/src/components/Button/Button.tsx

```diff
@@ -45,153 +48,118 @@ type CommonButtonProps<T> = {
-        );
-      }
+// Collect all the additional properties that one could supply to a button component
+// that will be passed down to whatever element or component is being used. This is
+// generally permissive in order to keep the typing simple at the expense of being
+// more accurate. `OtherProps` is generic so that we can extract any props from a
+// custom component that is being passed in. I'm trying to keep most of the complexity
+// in this section and leave the `ButtonProps` definition simpler. - PW
+//
+// Extend is a utility type that works like `Object.assign` where properties defined
+// on the latter type `N` override properties defined on the former type `M`
+type Extend<M, N> = Omit<M, Extract<keyof M, keyof N>> & N;
+type OtherProps<T extends ButtonComponentType> = Omit<
+  // Get all possible HTML attributes and override with any more specific props from
+  // the possibly custom component type `T`
+  Extend<React.HTMLAttributes<HTMLElement>, React.ComponentPropsWithRef<T>>,
+  // And omit any properties that we're defining on our own `CommonButtonProps`
+  keyof CommonButtonProps<T>
+>;
+
+export type ButtonProps<T extends ButtonComponentType> = CommonButtonProps<T> & OtherProps<T>;
+
+export const Button = <T extends ButtonComponentType>({
+  children,
+  className,
+  component,
+  disabled,
+  href,
+  inputRef,
+  inversed,
+  inverse,
+  onClick,
+  size,
+  variation,
+  type = 'button',
+  ...otherProps
+}: ButtonProps<T>) => {
+  if (process.env.NODE_ENV !== 'production') {
+    if (inverse) {
+      console.warn(
+        `[Deprecated]: Please remove the 'inverse' prop in <Button>, use 'inversed' instead. This prop has been renamed and will be removed in a future release.`
+      );
     }
-
-    this.handleClick = this.handleClick.bind(this);
-    this.handleKeyPress = this.handleKeyPress.bind(this);
-  }
-
-  // Get an object of props to pass to the rendered <Button> component
-  attrs(): any {
-    /**
-     * Since any number of arbitrary props can be passed into this component, we
-     * use a destructuring assignment to get only the props we want to pass to the
-     * rendered HTML element. For example, the "variation" prop is used to generate
-     * the classNames, but doesn't need to be passed to the rendered component, so
-     * we omit it here so that it's not included in the props object.
-     */
-    const {
-      className,
-      component,
-      inputRef,
-      inversed,
-      inverse,
-      onClick,
-      size,
-      variation,
-      ...props
-    } = this.props;
-
-    const attrs: any = {
-      className: this.classNames(),
-      ...props,
-    };
-
-    if (this.props.onClick) {
-      attrs.onClick = this.handleClick;
+    if (variation === 'danger') {
+      console.warn(
+        `[Deprecated]: Please remove the 'danger' variation prop in <Button>. This prop has will be removed in a future release.`
+      );
```

Comment:
Is this a good time to yank out these deprecated props?

Reply:
It would be, but we haven't discussed it prior to this change

Reply:
If we do, we should make it a separate PR